spec/lib/fields/base_spec.rb

```diff
@@ -167,4 +167,50 @@
       expect(field.required?).to eq(false)
     end
   end
+
+  describe "#data" do
+    context "when given nil data" do
+      it "reads the value from the resource" do
+        resource = double(attribute: "resource value")
+        field = field_class.new(:attribute, nil, :page, resource: resource)
+
+        expect(field.data).to eq("resource value")
+      end
+    end
+
+    context "when given non-nil data" do
+      it "uses the given data" do
+        resource = double(attribute: "resource value")
+        field = field_class.new(:attribute, "given value", :page, resource: resource)
+
+        expect(field.data).to eq("given value")
+      end
+    end
+
+    context "when given a :getter value" do
+      it "reads the attribute with the name of the value" do
+        resource = double(custom_getter: "custom value")
+        field = field_class.new(:attribute, :date, :page, resource: resource, getter: :custom_getter)
+
+        expect(field.data).to eq("custom value")
+      end
+    end
+
+    context "when given a :getter block" do
+      it "uses it to produce a value" do
+        resource = double("Model", custom_getter: "custom value")
+        field = field_class.new(:attribute, :date, :page, resource: resource, getter: ->(f) { f.resource.custom_getter + " from block" })
+
+        expect(field.data).to eq("custom value from block")
+      end
+    end
+
+    context "when given a :getter block" do
```

Comment:
This bit can be deleted as the two contexts are the same.

```suggestion
    end

    context "when given a :getter block" do
```

Was this part of my suggestion the other day? :flushed: